pestion/src/lib.rs

```diff
@@ -0,0 +1,443 @@
+
+    /// Converts this into [`Element`] which may imply parsing Pest syntax.
+    fn pest_to_element(&self) -> PestionResult<Self::Element>;
+}
+
+impl PestToElement for &str {
+    type Element = OwnedElement;
+
+    /// Parses a `str` slice as a Pest grammar and serializes the AST into [`Element`].
+    fn pest_to_element(&self) -> PestionResult<Self::Element> {
+        let pairs = PestParser::parse(Rule::grammar_rules, *self)?;
+        let ast = match consume_rules(pairs) {
+            Ok(ast) => ast,
+            Err(errors) => {
+                return if errors.is_empty() {
+                    invalid("Error converting Pest grammar to AST with no context")
+                } else {
+                    // TODO deal with more than one error..
+                    let err = errors.into_iter().next().unwrap();
+                    Err(err.into())
+                };
+            }
+        };
+
+        ast.pest_to_element()
+    }
+}
+
+impl PestToElement for Vec<AstRule> {
+    type Element = OwnedElement;
+
+    /// Converts a body of rules into a `struct` that has a rule for each field.
+    fn pest_to_element(&self) -> PestionResult<Self::Element> {
+        let mut fields = vec![];
+        for rule in self.iter() {
+            let rule_name = text_token(rule.name.clone());
+            let rule_value = rule.pest_to_element()?;
+            fields.push((rule_name, rule_value));
+        }
+        Ok(Self::Element::new_struct(fields))
+    }
+}
+
+impl PestToElement for AstRule {
+    type Element = OwnedElement;
+
+    /// Converts a Pest Rule into a `struct` that has the field for [`RuleType`] as a symbol
+    /// and a field for the [`Expr`].
+    fn pest_to_element(&self) -> PestionResult<Self::Element> {
+        let fields = vec![
+            (text_token("type"), self.ty.pest_to_element()?),
+            (text_token("expression"), self.expr.pest_to_element()?),
+        ];
+        Ok(Self::Element::new_struct(fields))
+    }
+}
+
+impl PestToElement for AstRuleType {
+    type Element = OwnedElement;
+
+    /// Serializes the enum into a symbolic value.
+    fn pest_to_element(&self) -> PestionResult<Self::Element> {
+        let sym_tok = text_token(match self {
+            RuleType::Normal => "normal",
+            RuleType::Silent => "silent",
+            RuleType::Atomic => "atomic",
+            RuleType::CompoundAtomic => "compound_atomic",
+            RuleType::NonAtomic => "non_atomic",
+        });
+
+        Ok(sym_tok.into())
+    }
+}
+
+impl PestToElement for Expr {
+    type Element = OwnedElement;
+
+    /// Generates a `sexp` representation of the rule expression.
+    fn pest_to_element(&self) -> PestionResult<Self::Element> {
+        use OwnedValue::*;
+
+        let element = Self::Element::new_sexp(match self.clone() {
+            Expr::Str(text) => vec![
+                text_token("string").into(),
+                text_token("exact").into(),
+                String(text).into(),
+            ],
```

Comment:
[`Element::new_sexp`]() takes an iterator:
```rust
    /// Builds a `sexp` from Sequence using Builder.
    fn new_sexp<I: IntoIterator<Item = Self::Element>>(seq: I) -> Self::Element;
```
so each of these `vec![]`s could be a slice.

Reply:
So just to be clear, I can rewrite this into an array that is transferred into an iterator, is that what you mean?

Reply:
Similarly, this is challenging because the slice iterator will yield `&OwnedElement` which is not compatible to `Iterator<Item = OwnedElement>`, I could rewrite this using `ArrayVec`, but I don't know if that is worth it.  What do you think about that, is there something I am missing with the slice suggestion?

Reply:
I have a factoring with `SmallVec`, could not do it with `ArrayVec` because of the limitations in that data type (specifically `Copy` requirements).

Reply:
Updated in [783de7f](https://github.com/partiql/partiql-lang-rust/pull/38/commits/783de7fae7c9f175f33af6e661a8c843cc49037a)